### PR TITLE
Add ability to load fonts one-by-one.

### PR DESCRIPTION
**Pull request by mmogilnitsky**
Example of usage:
https://jira.common.bluescape.com/browse/CORE-302
https://github.com/Bluescape/web-frontend/pull/8465

### CODE REVIEW

File: modules/canvaskit/canvaskit/types/index.d.ts [hunk 1]
```diff
@@ -1543,6 +1543,13 @@ export interface FontMgr extends EmbindObject<FontMgr> {
      * @param fontData
      */
     makeTypefaceFromData(fontData: ArrayBuffer): Typeface;
+
+    /**
+     * Add font to font manager.
+     * @param ArrayBuffer
+     * @param size - size of the buffer
+     */
+    addFont(fontData: ArrayBuffer, size: number): void;
```
[the remark this answers is not in the document]
Comment (mmogilnitsky) [reply modules/canvaskit/canvaskit/types/index.d.ts:1552]: Thanks!

File: modules/canvaskit/font.js [hunk 2]
```diff
@@ -127,15 +127,15 @@ CanvasKit._extraInitializations.push(function() {
   // arguments should all be arrayBuffers or be an array of arrayBuffers.
   CanvasKit.FontMgr.FromData = function() {
     if (!arguments.length) {
-      Debug('Could not make FontMgr from no font sources');
+      Debug('Make empty FontMgr from no font sources');
```
[the remark this answers is not in the document]
Comment (mmogilnitsky) [reply modules/canvaskit/font.js:130]: Ohh :(

File: modules/canvaskit/font.js [hunk 3]
```diff
@@ -156,6 +156,14 @@ CanvasKit._extraInitializations.push(function() {
     return fm;
   }
 
+  // fontData should an ArrayBuffer.
+  CanvasKit.FontMgr.prototype.addFont = function(fontData) {
+    var data = new Uint8Array(fontData);
+    var dataPtr = copy1dArray(data, 'HEAPU8');
+    this._addFont(dataPtr, data.byteLength);
+	  // The FontMgr has taken ownership of the bytes we allocated.
```
[the remark this answers is not in the document]
Comment (mmogilnitsky) [reply modules/canvaskit/font.js:164]: weird.

File: src/ports/SkFontMgr_custom.cpp [hunk 4]
```diff
@@ -185,6 +185,31 @@ SkFontMgr_Custom::SkFontMgr_Custom(const SystemFontLoader& loader) : fDefaultFam
...
+// after this operation. All doable, but ... too much of the API change for this
+// simple boot-strap solution.
+//
+// Another problem is that if something fails internally no information is
+// reported either. Same story, by the way with constructor - so no
+// improvement on that matter. There are at least two obvious reasons for the
+// failure memory overflow and bad buffer is suppiled. Bottom line problem
+// remains open.
+//
+// Even returning a newly added font family, if any, is questinable. Technically
```
[the remark this answers is not in the document]
Comment (mmogilnitsky) [reply src/ports/SkFontMgr_custom.cpp:198]: thanks

File: src/ports/SkFontMgr_custom.cpp [hunk 5]
```diff
@@ -185,6 +185,31 @@ SkFontMgr_Custom::SkFontMgr_Custom(const SystemFontLoader& loader) : fDefaultFam
...
+// this is doable, but ... it will be basically a hack. I.e. access the last
+// added family and fetch its Skia name. Just better not to.
+//
+// On a positive side for now we need only fontFamilyName. In fact we just need
+// the knowledge if anything got added and a list of all already loaded
+// fontFamilyNames. The result is to be acquired via standard API of:
+// countFamilies() and  getFamilyName(...) . A bit inefficient, but ... correct
+// in terms of getting the Skia's internal name.
+//
+// Note: it seems that Skia's intrnal naming is exactly the same as real name.
```
[the remark this answers is not in the document]
Comment (mmogilnitsky) [reply src/ports/SkFontMgr_custom.cpp:208]: thanks

File: src/ports/SkFontMgr_custom_embedded.cpp [hunk 6]
```diff
@@ -42,32 +42,30 @@ class EmbeddedSystemFontLoader : public SkFontMgr_Custom::SystemFontLoader {
...
-        int loaded = 0;
-        for (int i = 0; i < fNum; ++i) {
-            loaded += load_font_from_data(scanner, fDatas[i], fSizes[i], i, families);
-        }
+DataFontLoader::DataFontLoader(const uint8_t** datas, const size_t* sizes, int n)
+    : fDatas(datas)
+    , fSizes(sizes)
+    , fNum(n)
+{
+    printf("Here51 %zu %d\n", *sizes, n);
```
[the remark this answers is not in the document]
Comment (mmogilnitsky) [reply src/ports/SkFontMgr_custom_embedded.cpp:50]: thanks.

File: src/ports/SkFontMgr_custom_embedded.cpp [hunk 7]
```diff
@@ -42,32 +42,30 @@ class EmbeddedSystemFontLoader : public SkFontMgr_Custom::SystemFontLoader {
...
+    for (int i = 0; i < fNum; ++i) {
+        loaded += load_font_from_data(scanner, fDatas[i], fSizes[i], i, families);
     }
 
-private:
-    const uint8_t** fDatas;
-    const size_t* fSizes;
-    const int fNum;
-};
+    if (families->empty() && loaded) {
```
[the remark this answers is not in the document]
Comment (mmogilnitsky) [reply src/ports/SkFontMgr_custom_embedded.cpp:61]: And it is not being used anywhere. The access it to class locals only. These data can and should be derived from SkFontMgr directly. I guess it is some leftover of Skia development. I decided not to touch it. @ja2nicholl

File: src/ports/SkFontMgr_custom_embedded.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (mmogilnitsky) [reply src/ports/SkFontMgr_custom_embedded.cpp]: Sure :)

**Review by leoyanggit — APPROVED** [review]

**Review by ja2nicholl — APPROVED** [review]

**Pull request merged**